nat-lab/bin/core-api.py

```diff
@@ -0,0 +1,211 @@
+            public_key=req.public_key,
+            os=req.os,
+            os_version=req.os_version,
+            device_type=req.device_type,
+            nickname=req.nickname,
+            discovery_key=req.discovery_key,
+            relay_address=req.relay_address,
+            traffic_routing_supported=req.traffic_routing_supported,
+        )
+
+        self.server.add_machine(node)
+
+        self.server.send_notification()
+
+        pprint(node)
+        return node
+
+    def handle_machines_delete(self):
+        uid = self.path.removeprefix(self.machines_path)[1:]
+        print("uuid:", uid)
+
+        if self.server.remove_machine(uid):
+            self.server.send_notification()
+
+            print(f"{uid} removed")
+            self.send_response(204)
+        else:
+            print(f"{uid} unknown")
+            self.send_response(400)
+
+        self.end_headers()
+
+
+class CoreServer(HTTPServer):
+    # NOTE: In future this will most likely need to become a map from uuids to nodes so that it's possible
+    # to implement 'GET /v1/meshnet/machines/{machineIdentifier}/map'
+    known_machines_: set[str] = set()
+    mqttc = None
+
+    def send_notification(self):
+        message = {
+            "message": {
+                "data": {
+                    "metadata": {
+                        "message_id": str(uuid4()),
+                    },
+                    "event": {
+                        "attributes": {"affected_machines": list(self.known_machines_)}
+                    },
+                }
+            }
+        }
+
+        msg_info = self.mqttc.publish("meshnet", json.dumps(message), qos=1)
+        msg_info.wait_for_publish()
+
+    def add_machine(self, node):
+        self.known_machines_.add(node.identifier)
+
+    def remove_machine(self, uid):
+        if uid in self.known_machines_:
+            self.known_machines_.remove(uid)
+            return True
+        return False
+
+
+def run(mqttc, port=8080):
+    server_address = ("", port)
+    httpd = CoreServer(server_address, CoreApiHandler)
+    httpd.mqttc = mqttc
+    print("Starting httpd...")
+    httpd.serve_forever()
+
+
+def on_connect(_client, userdata, flags, reason_code, properties):
+    print(
+        f'Connected with result code userdata="{userdata}" flags="{flags}" reason_code="{reason_code}" properties="{properties}"'
+    )
+
+
+def on_message(_client, _userdata, msg):
+    print("got a message:", msg.topic + " " + str(msg.payload))
+
+
+def on_subscribe(_client, _userdata, mid, qos, _properties=None):
+    print(f"subscribed mid={mid}, qos={qos}")
+
+
+def main():
+    print("Starting core api simulator")
+    mqttc = mqtt.Client(
+        mqtt.CallbackAPIVersion.VERSION2, client_id="sender", protocol=mqtt.MQTTv311
+    )
+    mqttc.on_connect = on_connect
+    mqttc.on_message = on_message
+    mqttc.on_subscribe = on_subscribe
+    mqttc.connect("10.0.80.85", port=1883, keepalive=60)
```

Comment:
I think it would be nice to add some re-tries to connect (it might be the reason why core-api container failed to create in the pipeline, i.e. it cannot connect to the mqtt broker).

Reply:
Adding docker-compose dependency sounds good enough IMO